mlflow/pyfunc/__init__.py

```diff
@@ -266,7 +270,117 @@ def _load_model_env(path):
     return _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME).get(ENV, None)
 
 
-def load_model(model_uri, suppress_warnings=True):
+PyFuncOutput = Union[pandas.DataFrame, pandas.Series, np.ndarray, list]
+
+
+class PyFuncModel(object):
+    """
+    MLflow 'python function' model.
+
+    Wrapper around model implementation and metadata. This class is not meant to be constructed
+    directly. Instead, instances of this class are constructed and returned from
+    py:func:`mlflow.pyfunc.load_model`.
+
+    ``model_impl`` can be any Python object that implements the `Pyfunc interface
+    <https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#pyfunc-inference-api>`_, and is
+    by invoking the model's ``loader_module``.
+
+    ``model_meta`` contains model metadata loaded from the MLmodel file.
+    """
+
+    def __init__(self, model_meta: Model, model_impl: Any):
+        if not hasattr(model_impl, "predict"):
+            raise MlflowException("Model implementation is missing required predict method.")
+        self._model_meta = model_meta
+        self._model_impl = model_impl
+
+    def predict(self, data: pandas.DataFrame) -> PyFuncOutput:
+        """
+        Generate model predictions.
+        :param data: Model input as pandas.DataFrame.
+        :return: Model predictions as one of pandas.DataFrame, pandas.Series, numpy.ndarray or list.
+        """
+        input_schema = self._model_meta.get_input_schema()
+
+        if input_schema is not None:
+            def convert_type(name, values: pandas.Series, t: DataType):
+                if values.dtype == np.object and t not in (DataType.binary, DataType.string):
+                    values = values.infer_objects()
+
+                if values.dtype in (t.to_pandas(), t.to_numpy()):
```

Comment:
This is the short circuit in the case where the type is already appropriate, right? Maybe we can comment that.

Reply:
yup, I will add a comment.